README.md
# marbles

`marbles` is a Python `unittest` extension by [Jane Adams](mailto:jane@twosigma.com) and [Leif Walsh](mailto:leif@twosigma.com) that allows test authors to write richer tests that expose more information to test consumers on test failure. These information-rich tests help test consumers respond to failures without having to dig into the test code.

# Installing

The `marbles` namespace package contains two packages: `marbles.core` and `marbles.mixins`. `marbles.core` provides test cases that expose more information on test failure. `marbles.mixins` provides a set of semantically-rich assertions for use in both `marbles` and Python `unittest` tests. You can install the `marbles` package as a whole or install each subpackage individually.

You can install `marbles` directly with `setup.py` or using `pip`:

```bash
python setup.py install
# -or-
pip install .
```

You can also install `marbles` directly from GitLab:

```bash
pip install git+https://gitlab.twosigma.com/jane/marbles@<version tag>
```

You can also install `marbles` with conda:

```bash
conda install marbles
```

To install the subpackages individually, you can use `setup.py`:

```bash
cd marbles/core && python setup.py install
# -and/or-
cd marbles/mixins && python setup.py install
```

You can also install the subpackages with conda:

```bash
conda install marbles-core
# -and/or-
conda install marbles-mixins
```

Note that you cannot use `pip` to install subpackages individually or install subpackages individually from GitLab.

# Developing

Create an empty conda environment and install from the requirements.txt file. This will install dependencies and do a development ("editable") install of both of the subpackages.

```bash
conda create --yes --name marbles-dev python
source activate marbles-dev
pip install -r requirements.txt
```

Now you can edit code and run scripts or a Python shell. Imports of marbles modules and packages will come from the current directory, so your changes will take effect.

# Running Tests

Tests are written with the builtin `unittest` package, live in `tests/`, and are run with `setup.py`:

```bash
python setup.py test
```

## Test Coverage

You can get test coverage by running tests under `coverage`:

```bash
coverage run setup.py test
coverage report
```

Instead of `coverage report`'s text output you can get an HTML report including branch coverage information:

```bash
coverage run setup.py test
coverage html
cd build/coverage/html; python -m http.server 8080
```

and then visit [http://localhost:8080](http://localhost:8080).

# Building Documentation

Docs live in `docs/` and can be run with `setup.py`:

```bash
python setup.py build_sphinx
cd build/sphinx/html; python -m http.server 8080
```

and then visit [http://localhost:8080](http://localhost:8080).

# Releases

Version information is pulled from git tags by `versioneer`, so cutting a new release is as simple as tagging and pushing the tag:

```bash
git tag 0.6.9
git push origin 0.6.9
```

Once this tag exists, future package and documentation builds will automatically get that version, and users can `pip install` using that git tag from GitLab:

```bash
pip install git+https://gitlab.twosigma.com/jane/marbles@0.6.9
```

After tagging, you can release a new conda package with [Jenkins](http://tsmerejenkins.app.twosigma.com:8080/job/Datanomics/job/marbles/job/release/).

Other projects can depend on `marbles` using `dependency_links`:

```python
setup(
    ...,
    install_requires=['marbles==0.6.9'],
    dependency_links=['https://gitlab.twosigma.com/jane/marbles@0.6.9#egg=marbles-0.6.9'],
    ...
)
```

# Bugs

Please [open an issue](https://gitlab.twosigma.com/jane/marbles/issues/new?issue) to report any bugs.


docs/changelog.rst
=========
Changelog
=========

* :release:`0.12.3 <2023-12-20>`
* :bug:`128` Properly handle assertion failures during setUp/tearDown
* :support:`145` Support Python 3.12 (and drop 3.8)
* :release:`0.12.2 <2023-12-20>`
* :support:`144` Support Python 3.11
* :support:`139` Support Python 3.10
* :support:`138` Test on Windows
* :release:`0.12.1 <2020-08-02>`
* :support:`125` Fix long_description for PyPI
* :release:`0.12.0 <2020-05-23>`
* :support:`120` Upgrade to python 3.8
* :support:`119` Upgrade to pandas 1.0
* :feature:`50` Add a distutils command for marbles
* :release:`0.11.0 <2019-11-17>`
* :feature:`107` Allow triple-quoted annotations to be indented in source
* :support:`105` Fixed ``UniqueMixins`` literalinclude line numbers in docs
* :support:`101` Added note about how to execute logging configured tests
* :support:`99` Fixed ``assertCategoricalLevel(Not)In`` docstring parameters
* :release:`0.10.0 <2018-09-23>`
* :feature:`92` Improve indentation of multiline locals
* :support:`90` Added support for python 3.7
* :support:`88` Document how to install with conda
* :release:`0.9.5 <2018-06-24>`
* :support:`80` Added support for ``pandas<0.24``
* :bug:`58` Fixed test failure on OSX
* :release:`0.9.4 <2018-06-03>`
* :bug:`65` Fixed sdist installation
* :release:`0.9.3 <2018-06-02>`
* :support:`43` Added version bumping automation and maintenance documentation
* :support:`39` Added issue templates
* :bug:`40` Fixed "Locals" section in failure output to be hidden when
  no locals will be displayed
* :bug:`41` Removed developer dependencies in ``setup_requires`` and
  ``tests_require``
* :release:`0.9.2 <2018-05-19>`
* :support:`0`

  .. note:: First public release

* :support:`31` Added PyPI packaging
* :support:`28` Added Travis CI integration
* :support:`26` Added development automation and CI with tox
* :support:`17` Changed to pipenv for development environment management
* :support:`16` Added Contributor License Agreement forms
* :bug:`15` Added Creative Commons attribution for test content from Wikipedia
* :bug:`5` Changed copyright headers to refer to TSOS and the MIT license
* :bug:`1` Fixed tests to run in virtualenvs
* :support:`14` Removed DataFrame and Panel mixins
* :support:`18` Removed TS internal details from README
* :support:`21` Removed TS internal details from documentation and comments
* :support:`30` Removed TS internal conda recipe
* :release:`0.8.0 <2018-05-18>`
* :feature:`0` Large refactor and doc rewrite to prepare for open source
* :feature:`0` Added main method to provide ``python -m marbles``
* :support:`0` Split package into ``marbles.core`` and ``marbles.mixins``
* :feature:`0` Removed Traceback display for marbles assertion failures
* :feature:`0` Changed annotation to be optional with ``marbles.core.TestCase``
* :feature:`0` Changed test case and test method to log separately,
  and added marbles version
* :release:`0.6.9 <2017-10-18>`
* :support:`0`

  .. admonition:: Nice

     Nice

* :support:`0` Added conda recipe (internal only)
* :bug:`0` Fixed mixins that expect a specific type to raise
  ``TypeError`` instead of ``AssertionError``
* :bug:`0` Fixed source code extraction to find it inside eggs
* :release:`0.6.0 <2017-08-10>`
* :feature:`0` Added verbose logging option
* :feature:`0` Changed annotation wrapping to wrap paragraphs in
  annotations individually for better formatting
* :release:`0.5.0 <2017-03-20>`
* :feature:`0` Added mixins library
* :feature:`0` Added documentation about authoring good marbles docs
* :release:`0.4.0 <2017-02-28>`
* :feature:`0` Added richer text formatting in annotations
* :release:`0.3.0 <2017-02-23>`
* :feature:`0` Added assertion logging
* :release:`0.2.0 <2016-12-14>`
* :bug:`0` Fixed positional argument handling
* :feature:`0` Removed display of "private" locals
* :feature:`0` Removed extra ``message`` annotation
* :release:`0.1.0 <2016-10-19>`
* :feature:`0` Added annotation support in ``assert*`` methods
* :feature:`0` Added source code for the whole statement that failed
  to failure messages
* :feature:`0` Added ability to capture and display locals


noxfile.py
from concurrent.futures import ThreadPoolExecutor
from functools import wraps
import os
from pathlib import Path
import shutil
import sys

import nox

# Environment defaults: we want strict behavior in CI, and good local defaults
# so `nox` works reasonably and fast.

if os.environ.get('GITHUB_ACTIONS', False):
    nox.options.error_on_missing_interpreters = True
    nox.options.error_on_external_run = True
else:
    # local development defaults
    nox.options.sessions = ['test', 'flake8']
    nox.options.reuse_existing_virtualenvs = True

# The versions we test against, and the lowest version of Python we support.
# We should always update with pip-compile using the lowest version of Python
# so that we get all backport packages we'll need for that.

SUPPORTED_PYTHONS = ('3.9', '3.10', '3.11', '3.12')
MIN_SUPPORTED_PYTHON = '3.9'


# Helper session for integrating well with pip-sync.

def sync_session(*envs, install_marbles=False, **kwargs):
    '''Nox session decorator that uses pip-sync to manage dependencies.'''
    def decorator(f):
        @nox.session(**kwargs)
        @wraps(f)
        def wrapper(session: nox.Session, *args, **kwargs):
            session.install('pip-tools')
            concrete_reqs = [f'requirements/{env}.txt' for env in envs]
            if install_marbles and sys.platform == 'win32':
                concrete_reqs.append('requirements/windows.txt')
            session.run_always('pip-sync', *concrete_reqs)

            if install_marbles:
                session.install('-e', 'marbles/core', '-e', 'marbles/mixins',
                                '--no-deps')
            return f(session, *args, **kwargs)

        return wrapper
    return decorator


# Most useful testing code below: tests, coverage, linting, docs.

@sync_session('base', python=SUPPORTED_PYTHONS, install_marbles=True)
def test(session: nox.Session):
    '''Run tests, without coverage.'''
    for subpackage in ('marbles/core', 'marbles/mixins'):
        with session.chdir(subpackage):
            session.run('python', '-m', 'unittest', 'discover', 'tests')


@sync_session('coverage', python=SUPPORTED_PYTHONS, install_marbles=True,
              tags=['actions-test'])
def coverage(session: nox.Session):
    '''Run tests, with coverage.'''
    session.run('coverage', 'erase')

    with session.chdir('marbles/core'):
        session.run('coverage', 'run', '-m', 'unittest', 'discover', 'tests')
    examples_dir = Path('marbles/core/example_packages')
    for f in examples_dir.glob('**/*.coverage*'):
        shutil.move(str(f), 'marbles/core')

    with session.chdir('marbles/mixins'):
        session.run('coverage', 'run', '-m', 'unittest', 'discover', 'tests')

    session.run('coverage', 'combine', 'marbles/core',
                'marbles/mixins/.coverage', '.')
    session.run('coverage', 'report')
    session.run('coverage', 'html')
    session.run('coverage', 'xml')


@nox.session
def serve_coverage(session: nox.Session):
    '''Serve coverage data.'''
    with session.chdir('build/coverage/html'):
        session.run('python', '-m', 'http.server', *session.posargs)


@sync_session('lint', tags=['actions-check'])
def flake8(session: nox.Session):
    '''Lint with flake8.'''
    session.run('flake8')


@sync_session('docs', install_marbles=True, tags=['actions-check'])
def docs(session: nox.Session):
    '''Build docs with Sphinx.'''
    session.run('sphinx-build', '-b', 'html', '-WEa',
                '-d', 'build/sphinx/doctrees',
                'docs', 'build/sphinx/html')


@nox.session
def serve_docs(session: nox.Session):
    '''Serve built docs.'''
    with session.chdir('build/sphinx/html'):
        session.run('python', '-m', 'http.server', *session.posargs)


@sync_session('package', tags=['actions-check'])
def package(session: nox.Session):
    '''Build source and wheel distributions.'''
    for d in ('dist', 'marbles/core/dist', 'marbles/mixins/dist'):
        shutil.rmtree(d, ignore_errors=True)
    for d in ('.', 'marbles/core', 'marbles/mixins'):
        session.run('python', '-m', 'build', d)


# Administrative sessions: uploading to PyPI, updating dependencies, increasing
# the version strings.

@nox.session
def upload(session: nox.Session):
    '''Upload distributions to PyPI.'''
    session.install('twine')
    session.run(
        'twine', 'upload',
        'dist/*', 'marbles/core/dist/*', 'marbles/mixins/dist/*',
    )


@nox.session(python=MIN_SUPPORTED_PYTHON)
def pip_compile(session: nox.Session):
    '''Refresh concrete dependencies using pip-compile.

    You can upgrade our pinned dependencies by running ``nox -s pip_compile --
    -U``.
    '''
    session.install('pip-tools')
    args = ['--generate-hashes']

    for subpackage in ('marbles/core', 'marbles/mixins'):
        with session.chdir(subpackage):
            session.run('pip-compile', *args, *session.posargs)

    reqs_dir = Path('requirements')
    infiles = [
        req for req in reqs_dir.glob('*.in')
        if req.stem != 'windows' or sys.platform == 'win32'
    ]
    with ThreadPoolExecutor() as executor:
        executor.map(
            lambda infile: session.run(
                'pip-compile', *args, *session.posargs, str(infile)
            ),
            infiles
        )

    # Remove after https://github.com/jazzband/pip-tools/pull/1650
    for f in reqs_dir.glob('*.txt'):
        content = f.read_text()
        content = content.replace(f'{Path.cwd()}/', '')
        f.write_text(content)


@nox.session
def bumpversion(session: nox.Session):
    '''Update the version number: pass "major", "minor", or "patch".

    Use with positional arguments, e.g. ``nox -s bumpversion -- minor``.
    '''
    session.install('bump2version')
    session.run('bumpversion', *session.posargs)
